evennia/contrib/rpg/rpsystem/rpsystem.py

```diff
@@ -1333,30 +1336,30 @@ def get_search_result(
         """
         # we also want to use the default search method
         search_obj = super().get_search_result
-        is_builder = self.locks.check_lockstring(self, "perm(Builder)")
+        is_builder = self.permissions.check("Builder")
+        results = []
 
-        if candidates:
-            candidates = parse_sdescs_and_recogs(
+        if candidates is not None:
+            searched_results = parse_sdescs_and_recogs(
                 self, candidates, _PREFIX + searchdata, search_mode=True
             )
-            results = []
-            for candidate in candidates:
-                # we search by candidate keys here; this allows full error
-                # management and use of all kwargs - we will use searchdata
-                # in eventual error reporting later (not their keys). Doing
-                # it like this e.g. allows for use of the typeclass kwarg
-                # limiter.
-                results.extend(
-                    [obj for obj in search_obj(candidate.key, **kwargs) if obj not in results]
-                )
-
-            if not results and is_builder:
-                # builders get to do a global search by key+alias
-                results = search_obj(searchdata, **kwargs)
+            if not searched_results and is_builder:
+                # builders get to do a search by key
+                results = search_obj(searchdata, candidates=candidates, **kwargs)
+            else:
+                # we do a default search on each result by key, here, to apply extra filtering kwargs
+                for searched_obj in searched_results:
+                    results.extend(
+                        [
+                            obj
+                            for obj in search_obj(
+                                searched_obj.key, candidates=[searched_obj], **kwargs
+                            )
+                            if obj not in results
+                        ]
+                    )
         else:
-            # global searches with #drefs end up here. Global searches are
-            # only done in code, so is controlled, #dbrefs are turned off
-            # for non-Builders.
+            # no candidates means it's a global search, so we pass it back to the default
```

Comment:
there is no actual dbref check anywhere, this if/else is *only* checking if candidates were set and going to a global search if so. the dbref search itself is handled by the builder-permission fallback and the base search method.